File: Mapper/MaterialCategoryMapper.cs
using EFGetStarted.Model.DTO;
using EFGetStarted.Model.Entity;

namespace EFGetStarted.Mapper;

public class MaterialCategoryMapper : GenericMapper<MaterialCategory, MaterialCategoryPostDto, MaterialCategoryPutDto, MaterialCategoryGetDto>
{
    public override MaterialCategory ToEntity(MaterialCategoryPostDto dto)
    {
        return new MaterialCategory
        {
            Name = dto.Name
        };
    }
    
    public override MaterialCategory ToEntity(MaterialCategoryPutDto dto)
    {
        return new MaterialCategory
        {
            Name = dto.Name
        }.Let( it =>
        {
            if (dto.Id != null) it.Id = (int)dto.Id;
            return it;
        });
    }
    
    public override MaterialCategoryGetDto ToGetDto(MaterialCategory entity)
    {
        return new MaterialCategoryGetDto()
        {
            Id = entity.Id,
            Name = entity.Name
        };
    }
}

File: src/main/kotlin/com/example/recipecollection/mapper/MaterialCategoryMapper.kt
package com.example.recipecollection.mapper

import com.example.recipecollection.domain.MaterialCategory
import com.example.recipecollection.dto.MaterialCategoryDto
import com.example.recipecollection.dto.MaterialCategoryRequest
import org.mapstruct.Mapper
import org.mapstruct.MappingTarget

@Mapper(componentModel = "spring")
interface MaterialCategoryMapper {
    fun toDto(entity: MaterialCategory): MaterialCategoryDto
    fun toEntity(request: MaterialCategoryRequest): MaterialCategory
    fun updateEntity(request: MaterialCategoryRequest, @MappingTarget entity: MaterialCategory)
}
